src/vos/evtree.c

```diff
@@ -2522,11 +2540,144 @@ evt_ssof_adjust(struct evt_context *tcx, struct evt_node *nd,
 	return offset;
 }
 
+/**
+ * Sorted by Start Offset (SSOF)
+ *
+ * Extents are sorted by start offset first, then high to low epoch, then end
+ * offset
+ */
+
+/** Rectangle comparison for sorting */
+static int
+evt_ssof_cmp_rect(struct evt_context *tcx, const struct evt_rect *mbr,
+		  const struct evt_rect *rt1, const struct evt_rect *rt2)
+{
+	return evt_rect_cmp(rt1, rt2);
+}
+
+static int
+evt_ssof_insert(struct evt_context *tcx, struct evt_node *nd,
+		umem_off_t in_off, const struct evt_entry_in *ent)
+{
+	return evt_common_insert(tcx, nd, in_off, ent, evt_ssof_cmp_rect);
+}
+
+static int
+evt_ssof_adjust(struct evt_context *tcx, struct evt_node *nd,
+		struct evt_node_entry *ne, int at)
+{
+	return evt_common_adjust(tcx, nd, ne, at, evt_ssof_cmp_rect);
+}
+
 static struct evt_policy_ops evt_ssof_pol_ops = {
 	.po_insert		= evt_ssof_insert,
 	.po_adjust		= evt_ssof_adjust,
-	.po_split		= evt_ssof_split,
-	.po_rect_weight		= evt_ssof_rect_weight,
+	.po_split		= evt_even_split,
+	.po_rect_weight		= evt_common_rect_weight,
+};
+
+/**
+ * Sorted by distances to sides of bounding box
+ */
+
+/** Rectangle comparison for sorting */
+static int64_t
+evt_mbr_dist(const struct evt_rect *mbr, const struct evt_rect *rect)
+{
+	int64_t ldist = rect->rc_ex.ex_lo - mbr->rc_ex.ex_hi + 1;
+	int64_t rdist = mbr->rc_ex.ex_lo - rect->rc_ex.ex_hi + 1;
+
+	return ldist - rdist;
+}
+
+static int
+evt_sdist_cmp_rect(struct evt_context *tcx, const struct evt_rect *mbr,
+		   const struct evt_rect *rt1, const struct evt_rect *rt2)
+{
+	int64_t	dist1, dist2;
+
+	dist1 = evt_mbr_dist(mbr, rt1);
+	dist2 = evt_mbr_dist(mbr, rt2);
+
+	if (dist1 < dist2)
+		return -1;
+	if (dist1 > dist2)
+		return 1;
+
+	/* All else being equal, revert to ssof */
+	return evt_rect_cmp(rt1, rt2);
+}
+
+static int
+evt_sdist_split(struct evt_context *tcx, bool leaf, struct evt_node *nd_src,
+		struct evt_node *nd_dst)
+{
+	struct evt_node_entry	*entry_src;
+	struct evt_node_entry	*entry_dst;
+	int			 nr;
+	int64_t			 dist;
+
+	D_ASSERT(nd_src->tn_nr == tcx->tc_order);
+	nr = nd_src->tn_nr / 2;
+
+	nr += (nd_src->tn_nr % 2 != 0);
+
+	entry_src = evt_node_entry_at(tcx, nd_src, nr - 1);
+	dist = evt_mbr_dist(evt_node_mbr_get(tcx, nd_src), &entry_src->ne_rect);
+	if (dist > 0) {
+		do {
+			nr--;
+			if (nr == 1)
+				break;
+			entry_src = evt_node_entry_at(tcx, nd_src, nr - 1);
+			dist = evt_mbr_dist(evt_node_mbr_get(tcx, nd_src),
+					    &entry_src->ne_rect);
+		} while (dist > 0);
+	} else if (dist < 0) {
+		do {
+			nr++;
+			if (nr == nd_src->tn_nr - 1)
+				break;
+			entry_src = evt_node_entry_at(tcx, nd_src, nr - 1);
+			dist = evt_mbr_dist(evt_node_mbr_get(tcx, nd_src),
+					    &entry_src->ne_rect);
+		} while (dist < 0);
+	}
+	entry_src = evt_node_entry_at(tcx, nd_src, nr);
+	entry_dst = evt_node_entry_at(tcx, nd_dst, 0);
+	memcpy(entry_dst, entry_src, sizeof(*entry_dst) * (nd_src->tn_nr - nr));
```

Comment:
I have little concern on this uneven split, is it possible that split could fail in certain extreme case where all rectangles have same zero mbr distance? I think we'd better print some debug/warning log for far uneven split.

What about the performance comparison of EVT_FEAT_SORT_DIST vs. EVT_FEAT_SORT_DIST_EVEN?

Reply:
So far I've not been able to find any measurable performance impact.   It can happen though where we end up with an unbalanced tree with DIST.  I can add some debug.

I think I should probably put the bug fix in a separate patch.